### ISSUE TITLE
Calculate per-image metrics

### ISSUE THREAD

**Issue by lucalazzaroni** [ultralytics/ultralytics#18428]
Is it possible to calculate the metrics (precision, recall, mAP) per image? I trained a 5-class detection model and I need to extract the mAP50 for each test image. However, if I use the predict method, it can't be used with the labels, while if I use the val method, it seems to only work with datasets. I could create several 1-image test sets but I find it very inefficient and slow.
Could you help me in this?
Thank you in advance.

**Comment by Y-T-G** [comment]
You can't do that without modifying source code

**Comment by glenn-jocher** [comment]
@Y-T-G @Laughing-q yes this is a much needed feature actually. Otherwise it's impossible to correlate performance for anything other than class, when users really want to see performance correlated with image attributes like size, brightness, etc., and they also want to isolate the worst performing individual images to see if they can identify any correlations on their own, i.e. "the worst performing images are in the rain, etc."

**Comment by lucalazzaroni** [comment]
@glenn-jocher I totally agree with you. Do you have any idea on when this feature could be implemented? I'm trying to implement it by myself (I need only mAP50 atm) but it's very hard to put hands into the source code. I tried editing `val.py` and `metrics.py` but with no success.

**Comment by glenn-jocher** [comment]
Thank you for your efforts! At the moment, there isn't a specific timeline for this feature. However, feel free to share your progress or any implementation challenges as the community and Ultralytics team are always open to contributions or discussions. For guidance, you might want to explore how metrics are updated in `update_metrics()` within `val.py` [here](https://docs.ultralytics.com/reference/models/yolo/detect/val/).

**Comment by Laughing-q** [comment]
PR: https://github.com/ultralytics/ultralytics/pull/24089

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has now been merged in [PR #24089: `ultralytics 8.4.40` Per-image Precision and Recall](https://github.com/ultralytics/ultralytics/pull/24089).

### What changed
This PR adds **per-image validation metrics** to Ultralytics, which may resolve the issue by exposing image-level evaluation details instead of only dataset-wide averages. For supported tasks, validation results now include:

- `precision`
- `recall`
- `f1`
- `tp`
- `fp`
- `fn`

It also adds:
- a new `image_metrics` field in validation results
- support across detection, segmentation, pose, and OBB workflows
- multi-GPU gathering so per-image metrics are merged correctly across devices
- documentation updates with examples for accessing these results

This should make it much easier to inspect problematic images, debug labeling issues, and understand where validation performance is breaking down.

### Credits
Merged via [PR #24089](https://github.com/ultralytics/ultralytics/pull/24089) by @Laughing-q with contributions from @glenn-jocher and @Y-T-G.

### How to test
Since this PR was released in `8.4.40`, it can be tested with:

```bash
pip install -U ultralytics>=8.4.40
```

Or to test the latest code from `main`:

```bash
pip install git+https://github.com/ultralytics/ultralytics.git@main
```

If preferred, it is also fine to wait for the next official PyPI release if the updated package is not yet available in the current environment.

### Feedback
Please let us know whether these changes resolve the issue. Thanks 🙏 for reporting this, and any additional feedback is welcome if the issue still persists.
